File: ports/go/src/gui/kernel/bcsapplication.go
package kernel

import (
	"bcs/core/kernel"
	"fmt"
)

// BcsApplication manages the GUI application's control flow and main settings.
// Ported from C++ QCoreApplication/BcsApplication core
type BcsApplication struct {
	*kernel.BcsObject
	eventLoop *kernel.BcsEventLoop
	args      []string

	ApplicationName    string
	ApplicationVersion string
	OrganizationName   string
	OrganizationDomain string
}

// Global instance
var appInstance *BcsApplication

// NewBcsApplication initializes the GUI application state
func NewBcsApplication(args []string) *BcsApplication {
	app := &BcsApplication{
		BcsObject: kernel.NewBcsObject(nil),
		eventLoop: kernel.NewBcsEventLoop(),
		args:      args,
	}
	appInstance = app
	return app
}

// Instance returns the global application instance
func Instance() *BcsApplication {
	return appInstance
}

// Exec enters the main event loop and waits until exit() is called
func (app *BcsApplication) Exec() int {
	fmt.Println("BCS Application starting event loop...")
	app.eventLoop.Exec()
	return 0
}

// Quit requests the application to terminate
func (app *BcsApplication) Quit() {
	fmt.Println("BCS Application quitting...")
	app.eventLoop.Quit()
}

// ProcessEvents processes pending events in the queue
func (app *BcsApplication) ProcessEvents() {
	// Abstract wrapper matching processEvents
}

// PostEvent queues an event to a target receiver
func (app *BcsApplication) PostEvent(receiver *kernel.BcsObject, e *kernel.BcsEvent) {
	app.eventLoop.PostEvent(e)
}

// SendEvent directly dispatches an event to a receiver bypassing the queue
func (app *BcsApplication) SendEvent(receiver *kernel.BcsObject, e *kernel.BcsEvent) bool {
	if receiver != nil {
		return receiver.Event(e)
	}
	return false
}

File: ports/csharp/src/gui/kernel/BcsApplication.cs
using System;
using BCS.Core.Kernel;

namespace BCS.Gui.Kernel {
    // BcsApplication manages the GUI application's control flow
    public class BcsApplication : BcsObject {
        private BcsEventLoop _eventLoop;
        private string[] _args;

        public string ApplicationName { get; set; }
        public string ApplicationVersion { get; set; }
        public string OrganizationName { get; set; }
        public string OrganizationDomain { get; set; }

        public static BcsApplication Instance { get; private set; }

        public BcsApplication(string[] args) : base(null) {
            _eventLoop = new BcsEventLoop();
            _args = args;
            Instance = this;
        }

        public int Exec() {
            Console.WriteLine("BCS Application starting event loop...");
            _eventLoop.Exec();
            return 0;
        }

        public void Quit() {
            Console.WriteLine("BCS Application quitting...");
            _eventLoop.Quit();
        }

        public void ProcessEvents() {
            // Abstract wrapper matching processEvents
        }

        public void PostEvent(BcsObject receiver, BcsEvent e) {
            _eventLoop.PostEvent(e);
        }

        public bool SendEvent(BcsObject receiver, BcsEvent e) {
            if (receiver != null) {
                return receiver.Event(e);
            }
            return false;
        }
    }
}
